pkg/urbit/vere/saga.c

```diff
@@ -0,0 +1,730 @@
+  return 1;
+}
+
+//! @n (1) Arbitrarily choose 16 as the initial guess at the max number of
+//!        epochs.
+static c3_t
+_read_epoc_dirs(const c3_c* const dir_c, c3_c (**ent_c)[], size_t* ent_i)
+{
+  DIR* dir_u;
+  if ( !dir_c || !ent_c || !ent_i || !(dir_u = opendir(dir_c)) ) {
+    return 0;
+  }
+
+  *ent_c = NULL;
+  *ent_i = 0;
+
+  struct dirent* ent_u;
+  size_t         cap_i = 16; // (1)
+  c3_c(*dst_c)[dname_size]  = c3_malloc(cap_i * dname_size);
+  size_t dst_i         = 0;
+  while ( (ent_u = readdir(dir_u)) ) {
+    if ( !_is_epoc_dir(ent_u->d_name) ) {
+      continue;
+    }
+    if ( dst_i == cap_i ) {
+      cap_i *= 2;
+      dst_c = c3_realloc(dst_c, cap_i * dname_size);
+    }
+    strcpy(dst_c[dst_i++], ent_u->d_name);
+  }
+  if ( 0 == dst_i ) {
+    c3_free(dst_c);
+    return 0;
+  }
+  qsort(dst_c, dst_i, dname_size, _cmp_epocs);
+  *ent_c = dst_c;
+  *ent_i = dst_i;
+  return 1;
+}
+
+static inline void
+_remove_committed_events(u3_saga* const log_u)
+{
+  c3_list* eve_u = log_u->eve_u.lis_u;
+  size_t   len_i = log_u->eve_u.req_i;
+  for ( size_t idx_i = 0; idx_i < len_i; idx_i++ ) {
+    c3_free(c3_list_popf(eve_u));
+  }
+}
+
+//! @n (1) Attempt to commit events that were enqueued after the commit began.
+static void
+_uv_commit_after_cb(uv_work_t* req_u, c3_i sas_i)
+{
+  u3_saga* log_u = req_u->data;
+  log_u->act_t   = 0;
+
+  c3_t suc_t = log_u->asy_u.suc_t;
+  if ( suc_t ) {
+    _remove_committed_events(log_u);
+  }
+
+  c3_d las_d = u3_epoc_last_commit(log_u->epo_u.cur_u);
+  log_u->asy_u.com_f(log_u->asy_u.ptr_v, las_d, suc_t);
+
+  if ( UV_ECANCELED != sas_i ) { // (1)
+    u3_saga_commit(log_u, NULL, 0);
+  }
+}
+
+static void
+_uv_commit_cb(uv_work_t* req_u)
+{
+  u3_saga* log_u     = req_u->data;
+  u3_epoc* poc_u     = log_u->epo_u.cur_u;
+  c3_lode* nod_u     = c3_list_peekf(log_u->eve_u.lis_u);
+  size_t   len_i     = log_u->eve_u.req_i;
+  log_u->asy_u.suc_t = u3_epoc_commit(poc_u, nod_u, len_i);
+}
+
+static u3_epoc*
+_find_epoc(u3_saga* const log_u, const c3_d ide_d)
+{
+  c3_lode* nod_u = c3_list_peekb(log_u->epo_u.lis_u);
+  u3_epoc* poc_u;
+  while ( nod_u ) {
+    poc_u = c3_lode_data(nod_u);
+    if ( u3_epoc_has(poc_u, ide_d) ) {
+      break;
+    }
+    nod_u = c3_lode_prev(nod_u);
+  }
+  return nod_u ? poc_u : NULL;
+}
+
+//==============================================================================
+// Functions
+//==============================================================================
+
+//! @n (1) Persist metadata.
+//! @n (2) Create first epoch.
+u3_saga*
+u3_saga_new(const c3_path* const pax_u, const u3_meta* const met_u)
+{
+  u3_saga* log_u = c3_calloc(sizeof(*log_u));
+  if ( !(log_u->pax_u = c3_path_fp(pax_u)) ) {
+    goto free_event_log;
+  }
+  mkdir(c3_path_str(log_u->pax_u), 0700);
+
+  if ( !_create_metadata_files(log_u, met_u) ) { // (1)
+    goto free_event_log;
+  }
+
+  { // (2)
+    try_list(log_u->epo_u.lis_u = c3_list_init(), goto free_event_log);
```

Comment:
Omitting the `try_` would leak the implementation details of `c3_list_init()`. Creating a list is an operation that can logically fail, and it's in our best interest to not make any assumptions about how that failure occurs when using the `c3_list` interface.

Reply:
This feels like null-hypothesis juggling. `c3_list` is presumably part of the `c3` "package", and allocations in `c3` do not locally fail. Even if that counts as an "implementation detail", the question is: what should the `c3_list` API be? Specifying that it can fail when it actually cannot requires callers to include non-operational boilerplate, which further complicates already dense and subtle code.